### PR TITLE
Update @ariakit/react to 0.4.13

### PR DESCRIPTION
**Pull request by ciampo**
## What?

- Update `@ariakit/react` from `0.4.10` to `0.4.13`
- Update `@ariakit/test` from `0.4.2` to `0.4.5`
- Remove workarounds to `DropdownMenu` and `Composite` that are now provided directly by `ariakit`

## Why?

- Using the latest available version of ariakit allows to enjoy the latest features and bug fixes;
- Doing so frequently allows us to have leaner updates and fewer changes to test

## How?

- Updated the various `package.json` files, remove all references to the target dependencies
- Ran `npm cache clean --force && npm run distclean && npm install` to update the `package-lock.json` file
- Updated the various `package.json` files, re-adding the target dependencies with the latest versions
- Ran `npm cache clean --force && npm run distclean && npm install` to update the `package-lock.json` file
- Removed previous workarounds that are now fixed at the ariakit level

## Testing Instructions

- Smoke test ariakit-powered components Storybook and the editor: `Composite`, `Tabs`, `TabPanel`, `DropdownMenuV2`, `CustomSelectControl`, `Toolbar`, `Disclosure`, `Divider`, `RadioGroup`, `ToggleGroupControl`, `Tooltip`
- Smoke test dataviews UI using ariakit 

Looking at the release notes, here are the changes that could affect us:

- [Tab panels now support scroll restoration by default](https://ariakit.org/changelog#tab-panels-with-scroll-restoration). I think it's a nice addition, but if we want to be more conservative, we can either disable it completely on our version, or simply expose the prop to consumers of our component.
- [A bug fix](https://ariakit.org/changelog#other-updates-1) that makes the `data-focus-visible` attribute applied synchronously — which should fix a couple of bugs that we noticed in the past (https://github.com/WordPress/gutenberg/pull/64942, https://github.com/WordPress/gutenberg/pull/65387#discussion_r1788581144). I removed the custom workaround on our side.
- [`Composite`-based components have their elements tabbable if the id of the active composite item doesn't correspond to an element in the DOM](https://ariakit.org/changelog#accessible-composite-widgets-with-invalid-activeid). I removed the custom workaround added in https://github.com/WordPress/gutenberg/pull/65720

### Full changelog

<details>

<summary>`@ariakit/react` release notes</summary>

##### Accessible composite widgets with invalid `activeId`

We've improved the logic for composite widgets such as [Tabs](https://ariakit.org/components/tab) and [Toolbar](https://ariakit.org/components/toolbar) when the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state points to an element that is disabled or missing from the DOM. This can happen if an item is dynamically removed, disabled, or lazily rendered, potentially making the composite widget inaccessible to keyboard users.

Now, when the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state is invalid, all composite items will remain tabbable, enabling users to <kbd>Tab</kbd> into the composite widget. Once a composite item receives focus or the element referenced by the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state becomes available, the roving tabindex behavior is restored.

##### Other updates

-   Fixed regression in [`focusShift`](https://ariakit.org/reference/composite-provider#focusshift).
-   Fixed [Radio](https://ariakit.org/components/radio) to prevent `onChange` from triggering on radios that are already checked.
-   Fixed [`DisclosureContent`](https://ariakit.org/reference/disclosure-content) setting an incorrect `animating` state value during enter animations.
-   Improved JSDocs.
-   Updated dependencies: `@ariakit/react-core@0.4.13`

##### Tab panels with scroll restoration

Ariakit now supports scroll restoration for the [`TabPanel`](https://ariakit.org/reference/tab-panel) component. This allows you to control whether and how the scroll position is restored when switching tabs.

To enable scroll restoration, use the new [`scrollRestoration`](https://ariakit.org/reference/tab-panel#scrollrestoration) prop:

```jsx
// Restores the scroll position of the tab panel element when switching tabs
<TabPanel scrollRestoration />
```

By default, the scroll position is restored when switching tabs. You can set it to `"reset"` to return the scroll position to the top of the tab panel when changing tabs. Use the [`scrollElement`](https://ariakit.org/reference/tab-panel#scrollelement) prop to specify a different scrollable element:

```jsx
// Resets the scroll position of a different scrollable element
<div className="overflow-auto">
  <TabPanel
    scrollRestoration="reset"
    scrollElement={(panel) => panel.parentElement}
  />
</div>
```

##### Full height dialogs and on-screen virtual keyboards

A new [`--dialog-viewport-height`](https://ariakit.org/guide/styling#--dialog-viewport-height) CSS variable has been added to the [Dialog](https://ariakit.org/components/dialog) component. This variable exposes the height of the visual viewport, considering the space taken by virtual keyboards on mobile devices. Use this CSS variable when you have input fields in your dialog to ensure it always fits within the visual viewport:

```css
.dialog {
  max-height: var(--dialog-viewport-height, 100dvh);
}
```

##### Overriding composite state for specific methods

The [`next`](https://ariakit.org/reference/use-composite-store#next), [`previous`](https://ariakit.org/reference/use-composite-store#previous), [`up`](https://ariakit.org/reference/use-composite-store#up), and [`down`](https://ariakit.org/reference/use-composite-store#down) methods of the [composite store](https://ariakit.org/reference/use-composite-store) now accept an object as the first argument to override the composite state for that specific method. For example, you can pass a different [`activeId`](https://ariakit.org/reference/use-composite-store#activeId) value to the [`next`](https://ariakit.org/reference/use-composite-store#next) method so it returns the next item based on that value rather than the current active item in the composite store:

```js
const store = useCompositeStore({ defaultActiveId: "item1" });
const item3 = store.next({ activeId: "item2" });
```

It's important to note that the composite state is not modified when using this feature. The state passed to these methods is used solely for that specific method call.

##### Other updates

-   Fixed the ability to <kbd>Tab</kbd> out of a nested [Menu](https://ariakit.org/components/menu) within a modal [Dialog](https://ariakit.org/components/dialog).
-   Fixed CJS build on Next.js.
-   Enhanced performance on [Dialog](https://ariakit.org/components/dialog) backdrops.
-   Fixed [`Tab`](https://ariakit.org/reference/tab) to pass the [`rowId`](https://ariakit.org/reference/tab#rowid) prop when used with other composite widgets.
-   Improved JSDocs.
-   Updated dependencies: `@ariakit/react-core@0.4.12`

##### Tabs inside animated Combobox or Select

When rendering [Tab](https://ariakit.org/components/tab) inside [Combobox](https://ariakit.org/components/combobox) or [Select](https://ariakit.org/components/select), it now waits for the closing animation to finish before restoring the tab with the selected item. This should prevent an inconsistent UI where the tab is restored immediately while the content is still animating out. See [Select with Combobox and Tabs](https://ariakit.org/examples/select-combobox-tab).

##### Other updates

-   Updated [Combobox](https://ariakit.org/components/combobox) to immediately reset the [`activeId`](https://ariakit.org/reference/use-combobox-store#activeid) upon closing the popover.
-   Removed delay when applying the [`data-focus-visible`](https://ariakit.org/guide/styling#data-focus-visible) attribute.
-   Fixed mouse down on [`MenuButton`](https://ariakit.org/reference/menu-button) hiding the menu on Safari.
-   Improved JSDocs.
-   Updated dependencies: `@ariakit/react-core@0.4.11`

</details>

<details>

<summary>`@ariakit/test` release notes</summary>

#### @ariakit/test@0.4.5

- Updated dependencies: `@ariakit/core@0.4.12`

#### @ariakit/test@0.4.4

-   Fixed CJS build on Next.js.
-   Updated pointer events to initialize with a default `pointerType` value of `mouse`.
-   Updated dependencies: `@ariakit/core@0.4.11`

#### @ariakit/test@0.4.3

-   Updated dependencies: `@ariakit/core@0.4.10`

</details>

### CODE REVIEW

**Comment by ciampo** [comment]
After investigating the unit test failures, it seems like it could be a regression coming from ariakit. I've opened https://github.com/ariakit/ariakit/issues/4180 to confirm whether that's the case.

**Comment by ciampo** [comment]
> After investigating the unit test failures, it seems like it could be a regression coming from ariakit. I've opened [ariakit/ariakit#4180](https://github.com/ariakit/ariakit/issues/4180) to confirm whether that's the case.

That was indeed an ariakit regression that we caught on our end. 

A new version with the fix (`0.4.13`) was released, and I update this PR to use it.

There are still some `Tabs` unit tests failure, which I'm taking a look at right now. We have [a lot of custom logic](https://github.com/WordPress/gutenberg/blob/609a3c8d38a0608b41b1827ec4a8c1f812487ef2/packages/components/src/tabs/index.tsx#L72-L192) and maybe there's a chance to simplify it a bit

Otherwise, folks can already start to review the code changes and smoke test affected components

**Comment by diegohaz** [comment]
> editor/various/navigable-toolbar.spec.js:175:2 › Block Toolbar › Tab order of the block toolbar aligns with visual order

That's an interesting case to look into. Somehow the change to make composite items tabbable on invalid `activeId` is affecting the initial focus on the block toolbar when the screen size is reduced.

https://github.com/WordPress/gutenberg/blob/5c514368f47dbae9ebb722749cd85f64725c367e/test/e2e/specs/editor/various/navigable-toolbar.spec.js#L197-L203

It appears that `SlotFill` strikes again: when the toolbar is moved to the top, the buttons rendered via `BlockControls` are still receiving a stale toolbar store that points to elements from the floating toolbar, which are no longer connected to the DOM.

I think this highlights a broader issue with `SlotFill`, where `fillProps` aren't updating properly. When the toolbar is moved, if you inspect the toolbar store in the toolbar component itself and compare it with the toolbar store in a toolbar button inside `BlockControls`, they aren't the same object.

Adding something like `toolbarState.useState("activeId")` to the component below seems to resolve the issue, at least for this failing test. However, I don't believe this is the correct solution.

https://github.com/WordPress/gutenberg/blob/e89e988ba4a65b34357d5ee9ab78a8f6de73c35b/packages/block-editor/src/components/block-controls/slot.js#L21-L32

**Comment by ciampo** [comment]
@diegohaz it sounds like the correct fix would be to somehow trigger a re-render when the toolbar is moved from inline to fixed (and vice-versa)?

**Comment by tyxla** [comment]
> It appears that `SlotFill` strikes again: when the toolbar is moved to the top, the buttons rendered via `BlockControls` are still receiving a stale toolbar store that points to elements from the floating toolbar, which are no longer connected to the DOM.
> 
> I think this highlights a broader issue with `SlotFill`, where `fillProps` aren't updating properly. When the toolbar is moved, if you inspect the toolbar store in the toolbar component itself and compare it with the toolbar store in a toolbar button inside `BlockControls`, they aren't the same object.

I'm curious why it did't manifest before, while now it fails consistently. What might have changed in the latest couple of ariakit versions?

**Comment by diegohaz** [comment]
> I'm curious why it did't manifest before, while now it fails consistently. What might have changed in the latest couple of ariakit versions?

Here's my theory:

Ariakit now makes composite items (including toolbar items) tabbable when the `activeId` state is pointing to an element that's not connected to the DOM.

Those block controls apparently were already receiving a stale store, but since they weren't tabbable before, pressing Shift+Tab would move focus to the first item in the toolbar (which is not in a Slot/Fill). Somehow, this triggered a rerender on `BlockControlsSlot`, properly updating `fillProps` with the correct store.

I didn't test this thoroughly, but it's not making sense to me that BlockControlsSlot doesn't automatically rerenders when the store changes despite having `useContext(ToolbarContext)`.

**Comment by diegohaz** [comment]
https://github.com/ariakit/ariakit/pull/4202 may help with this specific failing test in the next version of Ariakit, but I believe there's still an underlying issue with slots or elements inside fills not receiving the correct context value, so this might just be sweeping the problem under the rug.

This is especially concerning to me:

> it's not making sense to me that BlockControlsSlot doesn't automatically rerender when the store changes despite having `useContext(ToolbarContext)`.

File: packages/dataviews/package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
 	"types": "build-types",
 	"sideEffects": false,
 	"dependencies": {
-		"@ariakit/react": "^0.4.10",
```
Comment (oandregal) [inline packages/dataviews/package.json:46]: For DataViews, I tested that list view's keyboard interactions still work as expected. Then, I also did some smoke testing of other places where I've seen it in use (modals, filters) and have nothing to report.

**Comment by jsnajdr** [comment]
I pushed a fix for the `SlotFill` issue. It was a bug in the `updateSlot` method that stores the latest `fillProps` prop passed to the `Slot` component into the slot/fill registry. When the toolbar moves from floating to sticky, there is a brief period of time when the old slot and new slot are still mounted, and the old slot updates the store with its old `fillProps` although a new slot is already registered. The `unregisterSlot` method already has a guard against this, checking if the current slot is still the registered one. Now I'm adding the same check also to `updateSlot` and it fixes the bug for me.

The `updateSlot` call is done on each render. The new version of Ariakit causes one more render of the old slot, that's all. That's enough to trigger the bug.

I'll file this fix also as a separate PR.

**Comment by ciampo** [comment]
I also worked on `Tabs` unit tests, which should hopefully not represent a blocker anymore.

As soon as I have some time, I'll rebase and see if there's anything else left before merging the update.

**Comment by ciampo** [comment]
I rebased to include the latest changes merged on `trunk`, let's see if tests pass.

**Comment by ciampo** [comment]
@WordPress/gutenberg-components This PR is now ready for a final review

**Comment by t-hamano** [comment]
> We've improved the logic for composite widgets such as [Tabs](https://ariakit.org/components/tab) and [Toolbar](https://ariakit.org/components/toolbar) when the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state points to an element that is disabled or missing from the DOM. This can happen if an item is dynamically removed, disabled, or lazily rendered, potentially making the composite widget inaccessible to keyboard users.
> 
> Now, when the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state is invalid, all composite items will remain tabbable, enabling users to <kbd>Tab</kbd> into the composite widget. Once a composite item receives focus or the element referenced by the [`activeId`](https://ariakit.org/reference/composite-provider#activeid) state becomes available, the roving tabindex behavior is restored.

I don't have enough knowledge about this, but is it possible to remove the workaround that exists in the `Tab` component as well?

https://github.com/WordPress/gutenberg/blob/8f6b37a5b71c911868d81c61a6f862a3109e2e9b/packages/components/src/tabs/tab.tsx#L32-L42

**Comment by ciampo** [comment]
That's a good point — I had added that workaround in https://github.com/WordPress/gutenberg/pull/66097 and forgot to remove it in this PR. I'm going to push a commit soon to remove that workaround.

@t-hamano — done.

**Review by tyxla — APPROVED** [review]
Did a bunch of smoke testing today and couldn't spot any issues. 

Nice to see a bunch of workarounds removed 💪 

I think this is good to ship 🚀

**Pull request merged**